### PR TITLE
Add stepper component

### LINKED ISSUE #219: Stepper Component
_Estimated Delivery Sprint - Sprint X_

- [x] Research
- [x] Specification
- [ ] Component and tests implementation
- [ ] API Documentation
- [x] Design implementation

### PR DESCRIPTION
**Pull request by viktorkombov**
Closes #219

### CODE REVIEW

**Comment by simeonoff** [comment]
I suggest we merge this PR without the implementation for variable size. We will then merge the theming package changes into master and implement sizing via CSS variables.

File: src/components/stepper/themes/step/light/step.base.scss [hunk 1]
```diff
@@ -0,0 +1,446 @@
...
+    margin-inline-start: calc(var(--body-indent--vertical) / 2);
+    padding-inline-start: var(--horizontal-body-padding, #{calc(var(--body-indent--vertical) / 2)});
+    padding-inline-end: var(--horizontal-body-padding, var(--vertical-body-padding));
+    padding-block-start: var(--horizontal-body-padding, var(--vertical-body-padding));
+    padding-block-end: var(--horizontal-body-padding, var(--vertical-body-padding));
+    z-index: 1;
+    position: relative;
+    color: color(gray, 900);
+
+    // Thous tow lines here are on purpose, they make sure that step body content
```
Comment (simeonoff) [inline src/components/stepper/themes/step/light/step.base.scss:173]:
```suggestion
    // Тhose two lines here are on purpose, they make sure that step body content
```

File: src/components/stepper/themes/step/light/step.base.scss [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+        inset-block-end: #{var(--step-separator-position)};
+    }
+
+    [part~='header'] {
+        align-items: var(--align-items, flex-start);
+        flex-direction: column-reverse;
+
+        &::before,
+        &::after {
+            inset-block-end: #{var(--step-separator-position)};
```
Comment (simeonoff) [inline src/components/stepper/themes/step/light/step.base.scss:208]: Why do we interpolate this here?

```suggestion
            inset-block-end: var(--step-separator-position);
```

File: src/components/stepper/themes/step/light/step.base.scss [hunk 3]
```diff
@@ -0,0 +1,446 @@
...
+        }
+    }
+}
+
+[part~='bottom'] {
+    align-items: var(--step-not-full-header-aligmnet, start);
+
+    &::before,
+    &::after {
+        inset-block-start: #{var(--step-separator-position)};
```
Comment (simeonoff) [inline src/components/stepper/themes/step/light/step.base.scss:218]:
```suggestion
        inset-block-start: var(--step-separator-position);
```

File: src/components/stepper/themes/step/light/step.base.scss [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+    }
+
+    [part~='header'] {
+        align-items: var(--align-items, flex-start);
+        flex-direction: column;
+
+        &::before,
+        &::after {
+            inset-block-start: #{var(--step-separator-position)};
```
Comment (simeonoff) [inline src/components/stepper/themes/step/light/step.base.scss:227]:
```suggestion
            inset-block-start: var(--step-separator-position);
```

File: src/components/stepper/themes/step/light/step.base.scss [hunk 5]
```diff
@@ -0,0 +1,446 @@
...
+
+        background: var(--header-background--focus);
+        box-shadow: inset 0 0 0 rem(1px) var(--header-shadow);
+    }
+}
+
+[part~='body'] {
+    display: var(--vertical-body-disply--last-of-type, var(--vertical-body-disply, none));
+    grid-row-start: var(--body-top, 2);
+    grid-column: span #{var(--steps-count)};
```
Comment (simeonoff) [inline src/components/stepper/themes/step/light/step.base.scss:163]:
```suggestion
    grid-column: span var(--steps-count);
```

File: src/components/stepper/themes/stepper/stepper.base.scss [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+// STEPPER [HORIZONTAL]
+:host([orientation='horizontal']) {
+    --horizontal-content-disply: block;
+    --vertical-separator-visibility: none;
+    --horizontal-body-padding: #{rem(8px)};
+
+    display: grid;
+    /* stylelint-disable */
+    grid-template-columns: minmax(var(--step-min-width--first-step), max-content) repeat(calc(#{var(--steps-count)} - 1), minmax(var(--step-min-width), auto));
```
Comment (simeonoff) [inline src/components/stepper/themes/stepper/stepper.base.scss:43]:
```suggestion
    grid-template-columns: minmax(var(--step-min-width--first-step), max-content) repeat(calc(var(--steps-count) - 1), minmax(var(--step-min-width), auto));
```

**Review by simeonoff — CHANGES_REQUESTED** [review]
Other than the few comments and suggestions I have, everything else LGTM.

File: src/components/stepper/step.ts [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+      this.dispatchEvent(
+        new CustomEvent('stepActiveChanged', { bubbles: true, detail: false })
+      );
+    }
+  }
+
+  @watch('disabled', { waitUntilFirstUpdate: true })
+  @watch('invalid', { waitUntilFirstUpdate: true })
+  @watch('optional', { waitUntilFirstUpdate: true })
+  protected disabledInvalidChange() {
```
Comment (igdmdimitrov) [inline src/components/stepper/step.ts:136]:
```suggestion
  protected disabledInvalidOptionalChange() {
```

File: src/components/stepper/step.ts [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+  }
+
+  @watch('complete', { waitUntilFirstUpdate: true })
+  protected completeChange() {
+    this.dispatchEvent(
+      new CustomEvent('stepCompleteChanged', { bubbles: true })
+    );
+  }
+
+  protected handleClick(event: MouseEvent): void {
```
Comment (igdmdimitrov) [inline src/components/stepper/step.ts:149]:
```suggestion
  private handleClick(event: MouseEvent): void {
```

File: src/components/stepper/step.ts [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+    event.stopPropagation();
+    if (this.isAccessible) {
+      this.dispatchEvent(
+        new CustomEvent('stepActiveChanged', { bubbles: true, detail: true })
+      );
+    }
+  }
+
+  protected handleKeydown(event: KeyboardEvent): void {
```
Comment (igdmdimitrov) [inline src/components/stepper/step.ts:158]:
```suggestion
  private handleKeydown(event: KeyboardEvent): void {
```

File: src/components/stepper/stepper.ts [hunk 10]
```diff
@@ -0,0 +1,450 @@
...
+        break;
+      }
+    }
+
+    return prevStep
+      ? prevStep
+      : this.steps.filter((step: IgcStepComponent) => step.isAccessible).pop();
+  }
+
+  private calculateLinearDisabledSteps(): void {
```
Comment (igdmdimitrov) [inline src/components/stepper/stepper.ts:354]:
```suggestion
  private updateStepsLinearDisabled(): void {
```

File: src/components/stepper/stepper.ts [hunk 11]
```diff
@@ -0,0 +1,450 @@
...
+      step.header?.setAttribute('aria-setsize', this.steps.length.toString());
+      step.header?.setAttribute('id', `igc-step-header-${index}`);
+      step.header?.setAttribute('aria-controls', `igc-step-content-${index}`);
+      if (index > 0) {
+        step.previousComplete = this.steps[index - 1].complete;
+      }
+    });
+  }
+
+  protected stepsChanged(): void {
```
Comment (igdmdimitrov) [inline src/components/stepper/stepper.ts:391]:
```suggestion
  private stepsChanged(): void {
```

**Review by igdmdimitrov — CHANGES_REQUESTED** [review]

**Comment by mmart1n** [comment]
> Some of the texts are not visible in dark variant.

@desig9stein could you please take a look at this?

**Comment by desig9stein** [comment]
> > Some of the texts are not visible in dark variant.
> 
> @desig9stein could you please take a look at this?

I don't see any problem with the texts. @igdmdimitrov Make sure that you have a dark background behind the stepper, otherwise, it is normal not to see the texts.

File: src/components/stepper/themes/stepper/stepper.base.scss (hunk 6 above)
Comment (desig9stein) [reply src/components/stepper/themes/stepper/stepper.base.scss]: For this one, my editor(webstorm) is complaining it gives me an error message "Mismatched parameters ([auto-fill | <integer>] , <line-names>+)" However, I removed it anyway and everything is working correctly, I will ignore that message from webstorm, they tend to have some issues regarding error messages for CSS/SCSS.

**Review by simeonoff — APPROVED** [review]

**Review by igdmdimitrov — APPROVED** [review]

**Pull request merged**